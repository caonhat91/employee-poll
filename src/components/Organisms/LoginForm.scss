.login-page {
    display: flex;
    flex-flow: column;
    align-items: center;
    justify-content: center;
    padding: 16px;
    gap: 64px;

    .h1 {
        width: 332px;
        height: 42px;

        /* H1 */
        font-family: var(--font-inter);
        font-style: normal;
        font-weight: 800;
        font-size: 32px;
        line-height: 39px;

        text-align: center;

        /* Linear Grotto Baby */
        background: linear-gradient(180deg, var(--baby-blue) 20%, var(--blue-grotto) 100%);
        -webkit-background-clip: text;
        -webkit-text-fill-color: transparent;
        background-clip: text;
        text-fill-color: transparent;
    }
}

.form {
    display: flex;
    flex-flow: column;
    align-items: flex-start;
    justify-content: center;
    gap: 24px;

    button,
    input[type=submit] {
        margin: 0 auto;
    }

    .control {
        display: flex;
        flex-flow: column;
        gap: 8px;
        min-height: 77px;

        label {
            font-size: 0.9em;
        }

        input {
            padding: 0.25rem 0.75rem;
            min-width: 280px;
            background: var(--white);
            border-radius: 0.25rem;
            border: 1px solid;
            border-color: var(--navy-blue);

            font-size: 1em;
            color: var(--navy-blue);

            transition: 300ms;

            &:focus {
                outline: 0;
                border-color: var(--navy-blue);
                box-shadow: 0px 0px 12px var(--blue-grotto);
            }

            &.invalid:invalid {
                border-color: var(--red);
                box-shadow: 0px 0px 12px var(--red);
            }
        }

        .error {
            color: var(--red);
            animation: fade 300ms linear;

            @keyframes fade {
                0% {
                    opacity: 0;
                }
            }
        }
    }
}