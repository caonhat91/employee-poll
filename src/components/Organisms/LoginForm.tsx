import { FormEvent, useState } from "react";

export type LoginType = {
    username: string,
    password: string
};

export default function LoginForm({ submit }: { submit: (_: LoginType) => void }) {

    const [frm, setFrm] = useState<LoginType>({
        username: '',
        password: ''
    });

    const changeFrm = (name: string, val: string) => {
        if (!name) {
            return;
        }
        setFrm({ ...frm, [name]: val });
    }

    const [errors, setErrors] = useState<LoginType | null>(null);

    const validateForm = () => {
        let newErrors: any = {};

        if (frm.username.trim() === '') {
            newErrors = { username: 'Username is required' };
        }

        if (frm.password.trim() === '') {
            newErrors.password = 'Password is required';
        }

        setErrors(newErrors);
        return Object.values(newErrors).length === 0;
    };

    const handleSubmit = (e: FormEvent) => {
        e.preventDefault();
        if (!validateForm()) {
            return;
        }
        submit(frm);
    }

    return (
        <form noValidate autoComplete="false" onSubmit={handleSubmit}>
            <label htmlFor="username">
                Username:
                <input type="text" name="username" id="username"
                    value={frm.username}
                    onChange={e => changeFrm(e.target.name, e.target.value)} />
            </label>
            {errors?.username && <p>{errors.username}</p>}
            <label htmlFor="password">
                Password:
                <input type="password" name="password" id="password"
                    value={frm.password}
                    onChange={e => changeFrm(e.target.name, e.target.value)} />
            </label>
            {errors?.password && <p>{errors.password}</p>}
            <button type="submit">Submit</button>
        </form>
    );
}