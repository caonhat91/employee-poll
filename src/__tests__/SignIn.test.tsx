import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import LoginForm, { LoginType } from '../components/Organisms/LoginForm';

describe("<LoginForm />", () => {
    test('renders login form', async () => {
        const handleSubmit = jest.fn((e: LoginType) => { });

        render(<LoginForm submit={handleSubmit} />)

        const inpUsername = await screen.findByLabelText('Username:');
        expect(inpUsername).toBeInTheDocument();

        const inpPassword = await screen.findByLabelText('Password:');
        expect(inpPassword).toBeInTheDocument();

        const btnSubmit = screen.getByText(/submit/i);
        expect(btnSubmit).toBeInTheDocument();
    });

    test('input username in form', async () => {
        const handleSubmit = jest.fn((e: LoginType) => { });

        render(<LoginForm submit={handleSubmit} />)

        const inpUsername = await screen.findByLabelText('Username:');
        fireEvent.change(inpUsername, { target: { value: 'test' } });

        expect(inpUsername.getAttribute('value')).toBe("test");
    });

    test('login form fail', async () => {
        const handleSubmit = jest.fn((e: LoginType) => { });

        render(<LoginForm submit={handleSubmit} />)

        // const inpUsername = await screen.findByLabelText('Username:');
        // fireEvent.change(inpUsername, { target: { value: '' } });

        const inpPassword = await screen.findByLabelText('Password:');
        fireEvent.change(inpPassword, { target: { value: 'password123' } });

        const btnSubmit = screen.getByText(/submit/i);
        fireEvent.click(btnSubmit);

        const usernameError = await screen.findByText(/Username is required/i);
        expect(usernameError).toBeInTheDocument();
    });

    test('login form success', async () => {
        let data: LoginType | undefined = undefined;
        const handleSubmit = jest.fn((e: LoginType) => data = e);

        render(<LoginForm submit={handleSubmit} />)

        const inpUsername = await screen.findByLabelText('Username:');
        fireEvent.change(inpUsername, { target: { value: 'sarahedo' } });

        const inpPassword = await screen.findByLabelText('Password:');
        fireEvent.change(inpPassword, { target: { value: 'password123' } });

        expect(data).toBeUndefined();

        const btnSubmit = screen.getByText(/submit/i);
        fireEvent.click(btnSubmit);

        expect(handleSubmit).toHaveBeenCalled();
        expect(data).toMatchObject({ username: 'sarahedo', password: 'password123' });
    });

    test("snapshot", () => {
        const handleSubmit = jest.fn((e: LoginType) => { });
        // eslint-disable-next-line testing-library/render-result-naming-convention
        const tree = render(<LoginForm submit={handleSubmit} />)
        expect(tree).toMatchSnapshot();
    });
});
