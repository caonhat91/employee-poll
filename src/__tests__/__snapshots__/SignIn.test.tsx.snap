// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<LoginForm /> snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="login-page"
      >
        <h1
          class="h1"
        >
          Login
        </h1>
        <form
          autocomplete="false"
          class="form"
          novalidate=""
        >
          <div
            class="control"
          >
            <label
              for="username"
            >
              Username:
            </label>
            <input
              id="username"
              name="username"
              required=""
              type="text"
              value=""
            />
          </div>
          <div
            class="control"
          >
            <label
              for="password"
            >
              Password:
            </label>
            <input
              id="password"
              name="password"
              required=""
              type="password"
              value=""
            />
          </div>
          <button
            type="submit"
          >
            Submit
          </button>
        </form>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="login-page"
    >
      <h1
        class="h1"
      >
        Login
      </h1>
      <form
        autocomplete="false"
        class="form"
        novalidate=""
      >
        <div
          class="control"
        >
          <label
            for="username"
          >
            Username:
          </label>
          <input
            id="username"
            name="username"
            required=""
            type="text"
            value=""
          />
        </div>
        <div
          class="control"
        >
          <label
            for="password"
          >
            Password:
          </label>
          <input
            id="password"
            name="password"
            required=""
            type="password"
            value=""
          />
        </div>
        <button
          type="submit"
        >
          Submit
        </button>
      </form>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
